src/ExpressionParser.php

```diff
@@ -51,9 +49,9 @@ class ExpressionParser
     public const OPERATOR_LEFT = 1;
     public const OPERATOR_RIGHT = 2;
 
-    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractUnary>}> */
+    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractExpression>}> */
     private array $unaryOperators;
-    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractBinary>, associativity: self::OPERATOR_*}> */
+    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractExpression>, associativity: self::OPERATOR_*}> */
```

Comment:
These 2 changes look wrong to me.

Reply:
Those values are set in construct
```
 $this->unaryOperators = $env->getUnaryOperators();
$this->binaryOperators = $env->getBinaryOperators();
```

They call the method on `$this->extensionSet`.
The ExtensionSet gets them from multiple ExtensionInterface, 
and getOperators is
```
/**
     * Returns a list of operators to add to the existing list.
     *
     * @return array<array> First array of unary operators, second array of binary operators
     *
     * @psalm-return array{
     *     array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractExpression>}>,
     *     array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class?: class-string<AbstractExpression>, associativity: ExpressionParser::OPERATOR_*}>
     * }
     */
    public function getOperators(): array;
```

This is because 
```
'??' => ['precedence' => 5, 'class' => NullCoalesceExpression::class, 'associativity' => ExpressionParser::OPERATOR_RIGHT],
```
is not an AbstractBinary but just an AbtractExpression.

Reply:
Indeed. I suppose all binary operators should extend AbstractBinary. I'm going to have a look at it.

Reply:
> Indeed. I suppose all binary operators should extend AbstractBinary. I'm going to have a look at it.

The issue already exists on 3.x if you want to take a look and fix it on the lower branch.
You can ping me after the fix, I'll update the PHPDoc if needed :)

Reply:
See #4482

Reply:
This PR can now be rebased.